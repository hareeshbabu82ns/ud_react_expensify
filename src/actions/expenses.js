import uuid from "uuid";
import databaseRef from "../firebase/firebase";

//-- Expenses Action Generators
export const addExpense = (expense = {}) => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return dispatch => {
    const {
      description = "",
      note = "",
      amount = 0,
      createdAt = Date.now()
    } = expenseData;
    const expense = { description, note, amount, createdAt };
    databaseRef
      .child("expenses")
      .push(expense)
      .then(ref => {
        console.log(ref);
        dispatch(addExpense({ id: ref.key, ...expense }));
      });
  };
};

export const removeExpense = ({ id = "" } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

export const editExpense = (id, payload) => ({
  type: "EDIT_EXPENSE",
  id,
  payload
});
