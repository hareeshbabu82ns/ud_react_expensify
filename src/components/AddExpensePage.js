import React from "react";

const AddExpensePage = () => (
  <div>
    <p>Add Expense</p>
  </div>
);

export default AddExpensePage;
