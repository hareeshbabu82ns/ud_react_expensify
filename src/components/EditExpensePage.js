import React from "react";

const EditExpensePage = props => (
  <div>
    <p>Edit Expense : {props.match.params.id}</p>
  </div>
);

export default EditExpensePage;
