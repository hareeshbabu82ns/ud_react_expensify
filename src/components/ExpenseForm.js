import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";
import "react-dates/lib/css/_datepicker.css";

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100.0).toString() : "",
      createdAt: moment(props.expense ? props.expense.createdAt : new Date()),
      createdAtFocused: false,
      error: ""
    };
  }
  onElementChange = e => {
    const st = {};
    st[e.target.name] = e.target.value;
    if (e.target.name === "amount") {
      //for special input validation
      if (e.target.value && !e.target.value.match(/^\d{1,}(\.\d{0,2})?$/))
        return;
    }
    this.setState(() => st);
  };
  onCreatedAtChanged = createdAt => {
    // console.log(createdAt.format("MMM Do, YYYY"));
    if (createdAt) this.setState(() => ({ createdAt }));
  };
  onCreatedAtFocused = ({ focused }) => {
    this.setState(() => ({ createdAtFocused: focused }));
  };
  onSubmit = e => {
    e.preventDefault();
    const error =
      this.state.description && this.state.amount
        ? ""
        : "Please provide Expense Description & Amount";
    this.setState(() => ({ error }));
    if (!error) {
      //prepare expense and propagate to parent
      const expense = {
        description: this.state.description,
        note: this.state.note,
        amount: Number(this.state.amount) * 100,
        createdAt: this.state.createdAt.valueOf()
      };
      this.props.onSubmit(expense);
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <div className="form-group">
            <input
              type="text"
              className="form-control"
              placeholder="Description"
              autoFocus
              value={this.state.description}
              onChange={this.onElementChange}
              name="description"
            />
          </div>
          <div className="form-group">
            <input
              type="text"
              className="form-control"
              placeholder="Amount"
              value={this.state.amount}
              onChange={this.onElementChange}
              name="amount"
            />
          </div>
          <div className="form-group">
            <SingleDatePicker
              className="form-control"
              date={this.state.createdAt}
              onDateChange={this.onCreatedAtChanged}
              focused={this.state.createdAtFocused}
              onFocusChange={this.onCreatedAtFocused}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
          <div className="form-group">
            <textarea
              className="form-control"
              placeholder="Notes"
              value={this.state.note}
              name="note"
              rows="5"
              onChange={this.onElementChange}
            />
          </div>
          <button className="btn btn-primary">Add Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
