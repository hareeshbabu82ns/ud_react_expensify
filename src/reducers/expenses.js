// -- Expenses Reducer
const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.expense];
    case "REMOVE_EXPENSE":
      return state.filter(({ id }) => id !== action.id);
    case "EDIT_EXPENSE":
      return state.map(item => {
        return item.id === action.id
          ? { ...item, ...action.payload, id: item.id }
          : item;
      });
    default:
      return state;
  }
};
