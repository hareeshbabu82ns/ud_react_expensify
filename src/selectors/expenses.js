import moment from "moment";

// filter expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter(expense => {
      const createdAt = moment(expense.createdAt);
      const startDateMatch = startDate
        ? moment(startDate).isSameOrBefore(createdAt, "day")
        : true;
      const endDateMatch = endDate
        ? moment(endDate).isSameOrAfter(createdAt, "day")
        : true;
      const textMatch = expense.description
        .toLowerCase()
        .includes(text.toLowerCase());

      return textMatch && startDateMatch && endDateMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt > b.createdAt ? -1 : 1;
      } else if (sortBy === "amount") {
        return a.amount > b.amount ? -1 : 1;
      }
    });
};

export const getExpensesTotal = expenses => {
  return expenses
    .map(e => e.amount)
    .reduce((amtPrev, amtCurr) => amtPrev + amtCurr);
};
