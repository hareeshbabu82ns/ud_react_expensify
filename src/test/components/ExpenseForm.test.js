import React from "react";
import { shallow } from "enzyme";
import moment from "moment";

import { ExpenseForm } from "../../components/ExpenseForm";
import expenses from "../fixtures/expenses";

test("should render ExpenseForm with default values", () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseForm with values", () => {
  const wrapper = shallow(<ExpenseForm expense={expenses[1]} />);
  expect(wrapper).toMatchSnapshot();
});

test("should throw error on submit", () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find("form").simulate("submit", { preventDefault: () => {} });
  expect(wrapper.state("error").length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test("should update state on description change", () => {
  const value = "some description";
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find("input[name='description']")
    .simulate("change", { target: { name: "description", value } });
  expect(wrapper.state("description")).toBe(value);
  expect(wrapper).toMatchSnapshot();
});

test("should update state on amount change", () => {
  const value = "33300";
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find("input[name='amount']")
    .simulate("change", { target: { name: "amount", value } });
  expect(wrapper.state("amount")).toBe(value);
  expect(wrapper).toMatchSnapshot();
});

test("should not update state on invalid amount change", () => {
  const value = "33ii300";
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find("input[name='amount']")
    .simulate("change", { target: { name: "amount", value } });
  expect(wrapper.state("amount")).not.toBe(value);
  expect(wrapper.state("amount")).toBe("");
  expect(wrapper).toMatchSnapshot();
});

test("should update date on change", () => {
  const now = moment();
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find("withStyles(SingleDatePicker)").prop("onDateChange")(now);
  expect(wrapper.state("createdAt")).toBe(now);
});

test("should call onSubmit with valid params", () => {
  const onSubmitSpy = jest.fn();
  const value = "some description";
  const wrapper = shallow(
    <ExpenseForm expense={expenses[1]} onSubmit={onSubmitSpy} />
  );
  wrapper
    .find("input[name='description']")
    .simulate("change", { target: { name: "description", value } });
  wrapper.find("form").simulate("submit", { preventDefault: () => {} });
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    description: value,
    note: expenses[1].note,
    amount: expenses[1].amount,
    createdAt: expenses[1].createdAt
  });
});
