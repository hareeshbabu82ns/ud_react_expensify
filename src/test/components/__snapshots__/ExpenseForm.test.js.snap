// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not update state on invalid amount change 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        name="description"
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        name="amount"
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <withStyles(SingleDatePicker)
        anchorDirection="left"
        customCloseIcon={null}
        customInputIcon={null}
        date={"1970-01-01T00:00:00.000Z"}
        daySize={39}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        id="date"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        verticalHeight={null}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        name="note"
        onChange={[Function]}
        placeholder="Notes"
        rows="5"
        value=""
      />
    </div>
    <button
      className="btn btn-primary"
    >
      Add Expense
    </button>
  </form>
</div>
`;

exports[`should render ExpenseForm with default values 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        name="description"
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        name="amount"
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <withStyles(SingleDatePicker)
        anchorDirection="left"
        customCloseIcon={null}
        customInputIcon={null}
        date={"1970-01-01T00:00:00.000Z"}
        daySize={39}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        id="date"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        verticalHeight={null}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        name="note"
        onChange={[Function]}
        placeholder="Notes"
        rows="5"
        value=""
      />
    </div>
    <button
      className="btn btn-primary"
    >
      Add Expense
    </button>
  </form>
</div>
`;

exports[`should render ExpenseForm with values 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        name="description"
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value="Electricity"
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        name="amount"
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value="50"
      />
    </div>
    <div
      className="form-group"
    >
      <withStyles(SingleDatePicker)
        anchorDirection="left"
        customCloseIcon={null}
        customInputIcon={null}
        date={"1970-01-01T00:00:00.000Z"}
        daySize={39}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        id="date"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        verticalHeight={null}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        name="note"
        onChange={[Function]}
        placeholder="Notes"
        rows="5"
        value=""
      />
    </div>
    <button
      className="btn btn-primary"
    >
      Add Expense
    </button>
  </form>
</div>
`;

exports[`should throw error on submit 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        name="description"
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        name="amount"
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <withStyles(SingleDatePicker)
        anchorDirection="left"
        customCloseIcon={null}
        customInputIcon={null}
        date={"1970-01-01T00:00:00.000Z"}
        daySize={39}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        id="date"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        verticalHeight={null}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        name="note"
        onChange={[Function]}
        placeholder="Notes"
        rows="5"
        value=""
      />
    </div>
    <button
      className="btn btn-primary"
    >
      Add Expense
    </button>
  </form>
</div>
`;

exports[`should throw error on submit 2`] = `
<div>
  <p>
    Please provide Expense Description & Amount
  </p>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        name="description"
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        name="amount"
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <withStyles(SingleDatePicker)
        anchorDirection="left"
        customCloseIcon={null}
        customInputIcon={null}
        date={"1970-01-01T00:00:00.000Z"}
        daySize={39}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        id="date"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        verticalHeight={null}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        name="note"
        onChange={[Function]}
        placeholder="Notes"
        rows="5"
        value=""
      />
    </div>
    <button
      className="btn btn-primary"
    >
      Add Expense
    </button>
  </form>
</div>
`;

exports[`should update state on amount change 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        name="description"
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        name="amount"
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value="33300"
      />
    </div>
    <div
      className="form-group"
    >
      <withStyles(SingleDatePicker)
        anchorDirection="left"
        customCloseIcon={null}
        customInputIcon={null}
        date={"1970-01-01T00:00:00.000Z"}
        daySize={39}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        id="date"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        verticalHeight={null}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        name="note"
        onChange={[Function]}
        placeholder="Notes"
        rows="5"
        value=""
      />
    </div>
    <button
      className="btn btn-primary"
    >
      Add Expense
    </button>
  </form>
</div>
`;

exports[`should update state on description change 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        autoFocus={true}
        className="form-control"
        name="description"
        onChange={[Function]}
        placeholder="Description"
        type="text"
        value="some description"
      />
    </div>
    <div
      className="form-group"
    >
      <input
        className="form-control"
        name="amount"
        onChange={[Function]}
        placeholder="Amount"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <withStyles(SingleDatePicker)
        anchorDirection="left"
        customCloseIcon={null}
        customInputIcon={null}
        date={"1970-01-01T00:00:00.000Z"}
        daySize={39}
        disabled={false}
        displayFormat={[Function]}
        enableOutsideDays={false}
        firstDayOfWeek={null}
        focused={false}
        hideKeyboardShortcutsPanel={false}
        horizontalMargin={0}
        id="date"
        initialVisibleMonth={null}
        inputIconPosition="before"
        isDayBlocked={[Function]}
        isDayHighlighted={[Function]}
        isOutsideRange={[Function]}
        isRTL={false}
        keepOpenOnDateSelect={false}
        monthFormat="MMMM YYYY"
        navNext={null}
        navPrev={null}
        numberOfMonths={1}
        onClose={[Function]}
        onDateChange={[Function]}
        onFocusChange={[Function]}
        onNextMonthClick={[Function]}
        onPrevMonthClick={[Function]}
        openDirection="down"
        orientation="horizontal"
        phrases={
          Object {
            "calendarLabel": "Calendar",
            "chooseAvailableDate": [Function],
            "clearDate": "Clear Date",
            "closeDatePicker": "Close",
            "dateIsUnavailable": [Function],
            "enterKey": "Enter key",
            "escape": "Escape key",
            "hideKeyboardShortcutsPanel": "Close the shortcuts panel",
            "homeEnd": "Home and end keys",
            "jumpToNextMonth": "Move forward to switch to the next month",
            "jumpToPrevMonth": "Move backward to switch to the previous month",
            "keyboardNavigationInstructions": "Press the down arrow key to interact with the calendar and
  select a date. Press the question mark key to get the keyboard shortcuts for changing dates.",
            "keyboardShortcuts": "Keyboard Shortcuts",
            "leftArrowRightArrow": "Right and left arrow keys",
            "moveFocusByOneDay": "Move backward (left) and forward (right) by one day",
            "moveFocusByOneMonth": "Switch months",
            "moveFocusByOneWeek": "Move backward (up) and forward (down) by one week",
            "moveFocustoStartAndEndOfWeek": "Go to the first or last day of a week",
            "openThisPanel": "Open this panel",
            "pageUpPageDown": "page up and page down keys",
            "questionMark": "Question mark",
            "returnFocusToInput": "Return to the date input field",
            "selectFocusedDate": "Select the date in focus",
            "showKeyboardShortcutsPanel": "Open the keyboard shortcuts panel",
            "upArrowDownArrow": "up and down arrow keys",
          }
        }
        placeholder="Date"
        readOnly={false}
        renderCalendarInfo={null}
        renderDay={null}
        renderMonth={null}
        reopenPickerOnClearDate={false}
        required={false}
        screenReaderInputMessage=""
        showClearDate={false}
        showDefaultInputIcon={false}
        verticalHeight={null}
        weekDayFormat="dd"
        withFullScreenPortal={false}
        withPortal={false}
      />
    </div>
    <div
      className="form-group"
    >
      <textarea
        className="form-control"
        name="note"
        onChange={[Function]}
        placeholder="Notes"
        rows="5"
        value=""
      />
    </div>
    <button
      className="btn btn-primary"
    >
      Add Expense
    </button>
  </form>
</div>
`;
